Refactors deselection code

steam opened the pull request:
Both `collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath)` and `collectionView(_ collectionView: UICollectionView, didDeselectItemAt indexPath: IndexPath)` now make use of the same deslection code.

When selecting an item ALL previously selected items that exceed the maximumSelection count are deselcted instead of only deselecting the first one.

Fixes a bug that required selecting an image multiple times after it had previously been selected and deselected.

steam committed:
Remove deselectAsset calls from didSelectItemAt

These callbacks cause a crash when the maximumSelection is 1 due to selectedAssets having the item at previousItemIndex removed in didDeselectItemAt down below.

colinta commented on ImagePickerSheetController/ImagePickerSheetController/ImagePickerSheetController.swift:
```diff
@@ -429,13 +429,8 @@ extension ImagePickerSheetController: UICollectionViewDelegate {
             if selectedAssetIndices.count >= maximumSelection,
                 let previousItemIndex = selectedAssetIndices.first {
-                    let deselectedAsset = selectedAssets[previousItemIndex]
-                    delegate?.controller?(self, willDeselectAsset: deselectedAsset)
```
Why is it unsafe/undesired to call the delegate methods here?  Line 433 is, effectively, deselecting the first selected item, and so notifying the delegate makes sense to me.

steam replied:
To me, it seems odd to call **de**select delegate functions called `willDeselectAsset` and `didDeselectAsset`in the `UICollectionViewDelegate` **select** function `didSelectItemAt`. They are already called in the `UICollectionViewDelegate` **de**select function `didDeselectItemAt `. 

The crash happens because `didDeselectItemAt` calls `selectedAssetIndices.remove(at: index)` which removes the item from `selectedAssets` causing `selectedAssets[previousItemIndex]` to crash with an index out of bounds exception.

steam replied:
The crash occurs when you set `maximumSelection` to 1 and then try to select a different image after selecting one already (all before dismissing the image picker sheet ).

steam committed:
Refactors deselection code

Both `collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath)` and `collectionView(_ collectionView: UICollectionView, didDeselectItemAt indexPath: IndexPath)` now make use of the same deslection code.

When selecting an item ALL previously selected items that exceed the maximumSelection count are deselcted instead of only deselecting the first one.

Fixes a bug that required selecting an image multiple times after it had previously been selected and deselected.
